### ISSUE TITLE
jOOQ Open Source Edition code generation fails on MySQL 5.7 and MariaDB 10.3

### ISSUE THREAD

**Issue by ahamana** [jOOQ/jOOQ#9832]
### Expected behavior and actual behavior:
When generating codes with jOOQ 3.13.0, the following exception occurs.
If adding FROM clause (from dual) before "where exists", MySQL returns not an error but '1' correctly.
\* select 1 as one from dual where exists (select 1 as one from information_schema.TABLES where (information_schema.TABLES.TABLE_NAME = 'proc' and information_schema.TABLES.TABLE_SCHEMA = 'mysql'));

```                              
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@  @@        @@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@        @@@@@@@@@@
@@@@@@@@@@@@@@@@  @@  @@    @@@@@@@@@@
@@@@@@@@@@  @@@@  @@  @@    @@@@@@@@@@
@@@@@@@@@@        @@        @@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@        @@        @@@@@@@@@@
@@@@@@@@@@    @@  @@  @@@@  @@@@@@@@@@
@@@@@@@@@@    @@  @@  @@@@  @@@@@@@@@@
@@@@@@@@@@        @@  @  @  @@@@@@@@@@
@@@@@@@@@@        @@        @@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@  @@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@  Thank you for using jOOQ 3.13.0
                                      
2月 13, 2020 9:27:26 午後 org.jooq.tools.JooqLogger info
情報: ARRAYs fetched           : 0 (0 included, 0 excluded)
2月 13, 2020 9:27:26 午後 org.jooq.tools.JooqLogger info
情報: Enums fetched            : 0 (0 included, 0 excluded)
2月 13, 2020 9:27:26 午後 org.jooq.tools.JooqLogger info
情報: Packages fetched         : 0 (0 included, 0 excluded)
2月 13, 2020 9:27:26 午後 org.jooq.tools.JooqLogger warn
警告: SQL exception            : Exception while executing meta query: (conn=6) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'where exists (select 1 as one from information_schema.TABLES where (information_' at line 1

If you think this is a bug in jOOQ, please report it here: https://github.com/jOOQ/jOOQ/issues/new

Note you can mute some exceptions using the configuration/onError flag

'''sql
select 1 as one
where exists (
  select 1 as one
  from information_schema.TABLES
  where (
    information_schema.TABLES.TABLE_NAME = 'proc'
    and information_schema.TABLES.TABLE_SCHEMA = 'mysql'
  )
)
'''

2月 13, 2020 9:27:26 午後 org.jooq.tools.JooqLogger error
重大: Cannot read C:\Users\ohba\git\koinomi\app\build\output\tmp\generateDataJooqSchemaSource\config.xml. Error : Error generating code for catalog 
org.jooq.codegen.GeneratorException: Error generating code for catalog 
	at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:399)
	at org.jooq.codegen.GenerationTool.run0(GenerationTool.java:845)
	at org.jooq.codegen.GenerationTool.run(GenerationTool.java:222)
	at org.jooq.codegen.GenerationTool.generate(GenerationTool.java:217)
	at org.jooq.codegen.GenerationTool.main(GenerationTool.java:189)
Caused by: java.lang.RuntimeException: org.jooq.exception.DataAccessException: SQL [select 1 as one where exists (select 1 as one from information_schema.TABLES where (information_schema.TABLES.TABLE_NAME = ? and information_schema.TABLES.TABLE_SCHEMA = ?))]; (conn=6) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'where exists (select 1 as one from information_schema.TABLES where (information_' at line 1
	at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:2545)
	at org.jooq.meta.AbstractDatabase.getRoutines(AbstractDatabase.java:2075)
	at org.jooq.codegen.JavaGenerator.generateSchemaIfEmpty(JavaGenerator.java:432)
	at org.jooq.codegen.JavaGenerator.generateCatalogIfEmpty(JavaGenerator.java:419)
	at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:393)
	... 4 more
Caused by: org.jooq.exception.DataAccessException: SQL [select 1 as one where exists (select 1 as one from information_schema.TABLES where (information_schema.TABLES.TABLE_NAME = ? and information_schema.TABLES.TABLE_SCHEMA = ?))]; (conn=6) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'where exists (select 1 as one from information_schema.TABLES where (information_' at line 1
	at org.jooq_3.13.0.MYSQL.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2751)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:755)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:385)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:502)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:471)
	at org.jooq.impl.AbstractResultQuery.fetchLazyNonAutoClosing(AbstractResultQuery.java:485)
	at org.jooq.impl.AbstractResultQuery.fetchOne(AbstractResultQuery.java:654)
	at org.jooq.impl.SelectImpl.fetchOne(SelectImpl.java:2884)
	at org.jooq.impl.DefaultDSLContext.fetchExists(DefaultDSLContext.java:4372)
	at org.jooq.impl.DefaultDSLContext.fetchExists(DefaultDSLContext.java:4382)
	at org.jooq.meta.AbstractDatabase.exists1(AbstractDatabase.java:516)
	at org.jooq.meta.mysql.MySQLDatabase.exists0(MySQLDatabase.java:630)
	at org.jooq.meta.AbstractDatabase.exists(AbstractDatabase.java:473)
	at org.jooq.meta.mysql.MySQLDatabase.is8(MySQLDatabase.java:255)
	at org.jooq.meta.mysql.MySQLDatabase.getRoutines0(MySQLDatabase.java:558)
	at org.jooq.meta.AbstractDatabase$15.run(AbstractDatabase.java:2078)
	at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:2535)
	... 8 more
Caused by: java.sql.SQLSyntaxErrorException: (conn=6) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'where exists (select 1 as one from information_schema.TABLES where (information_' at line 1
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:243)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.getException(ExceptionMapper.java:164)
	at org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:258)
	at org.mariadb.jdbc.ClientSidePreparedStatement.executeInternal(ClientSidePreparedStatement.java:225)
	at org.mariadb.jdbc.ClientSidePreparedStatement.execute(ClientSidePreparedStatement.java:145)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:209)
	at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet(Tools.java:3956)
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:295)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:371)
	... 22 more
Caused by: java.sql.SQLException: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'where exists (select 1 as one from information_schema.TABLES where (information_' at line 1
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.readErrorPacket(AbstractQueryProtocol.java:1598)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.readPacket(AbstractQueryProtocol.java:1460)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.getResult(AbstractQueryProtocol.java:1423)
	at org.mariadb.jdbc.internal.protocol.AbstractQueryProtocol.executeQuery(AbstractQueryProtocol.java:240)
	at org.mariadb.jdbc.ClientSidePreparedStatement.executeInternal(ClientSidePreparedStatement.java:216)
	... 27 more
```

### Versions:

- jOOQ: 3.13.0
- Java: Corretto-11.0.6.10.1 (build 11.0.6+10-LTS)
- Database (include vendor): MySQL 5.7
- OS: Windows 10
- JDBC Driver (include name if inofficial driver): MariaDB Connector/J 2.5.4

**Comment by lukaseder** [comment]
Thank you very much for your report. This regression was caused by an improvement #7421, where we stopped rendering the `DUAL` clause for MySQL 8+ and MariaDB 10.4+, as we found it is no longer necessary in these versions.

Of course, that decision was flawed, as it is currently not possible to easily configure the dialect used in the `jOOQ-meta` classes used for code generation.

As a workaround, you could subclass the `MySQLDatabase` class in your code generator configuration, and provide the `SQLDialect.MYSQL_5_7` dialect explicitly.

Do note that the jOOQ Open Source Edition does not formally support MySQL 5.7, only MySQL 8+. However, this bug also appears in the commercial jOOQ distributions, which is why we'll fix this immediately.

**Comment by ahamana** [comment]
@lukaseder Many thanks for the workaround.  I have understood that the jOOQ Open Source Edition does not formally support MySQL 5.7.

**Comment by lukaseder** [comment]
An additional workaround would be to override the `MySQLDatabase.exists0()` methods in your own implementation.

**Comment by lukaseder** [comment]
We have fixed this for the commercial distributions of jOOQ: https://github.com/jOOQ/jOOQ/issues/9834

While the jOOQ Open Source Edition does not formally support MySQL 5.7, I think this particular case is rather unfortunate, as it completely prevents using the jOOQ Open Source Edition with MySQL 5.7 (and MariaDB 10.3, for that matter).

We'll devise an additional workaround that prevents this problem in the jOOQ Open Source Edition.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.14.0. Backport scheduled for 3.13.1 (#9860). Will release 3.13.1 today (see https://github.com/jOOQ/jOOQ/milestone/181?closed=1)

**Comment by lukaseder** [comment]
Looking at this once more. The problem is the call to `DSLContext::fetchExists`

```java
    public boolean fetchExists(Select<?> query) {
        return selectOne().whereExists(query).fetchOne() != null;
    }
```

If we change that to:

```java
    public boolean fetchExists(Select<?> query) {
        return fetchValue(field(exists(query)));
    }
```

Then the queries run by the code generator will continue to run also on older versions of MySQL and MariaDB. At the same time, we can implement #7421 and move our generated SQL forward for those users who use newer versions of their RDBMS.

In the long run, we continue officially supporting older versions of various RDBMS only in our commercial distributions, see: https://www.jooq.org/download/support-matrix

So, the recommendation is to either:

- upgrade the RDBMS or
- upgrade the jOOQ price plan
- stay with a jOOQ version that officially supports the RDBMS version
